## Measuring multiomic fingerprints of wellness in a genetically tractable yeast population

Wellness is a new frontier in personalized medicine. Start-ups have recently emerged to quantify wellness and provide data-driven approaches to its management.  These projects combine DNA sequencing with longitudinal molecular profiling, as well as coaching to improve health and detect anomalies before they manifest disease. From a technical point of view, however, combining genetic information with molecular profiles to predict phenotype remains a significant challenge. Does integration across multiple scales of molecular information allow more sensitive detection of changes in health? Can many weakly predictive genetic markers be combined with real-time molecular profiling to identify personalized molecular intervention points? As an EIPOD fellow I will address these questions by quantifying the relationship between genetic, molecular, and environmental factors in a genetically tractable yeast population for a phenotype related to wellness, aging.

Yeast is a well established model of aging in eukaryotes. A yeastâ€™s longevity can be quantified as replicative life span (RLS), or the number of daughters produced before irreversible cell cycle arrest. High-throughput platforms have been established to measure RLS by isolating a single mother cell using microscopy and microfluidic devices [1]. As a first step in my project, I will establish an experimental system to measure RLS in a genetically diverse yeast population generated by the Steinmetz lab.  Each of 140 haploid yeast segregants from a cross between the laboratory strain of S. cerevisiae, S288c, and a pathogenic strain isolated from the lungs of an AIDS patient, YJM789, will be characterized for RLS. A recent study identified SIR2 as a primary regulator of lifespan in yeast [2]. I will improve these measurements by using more sensitive QTL detection methods, a more finely measured phenotype, and segregants with greater genomic resolution to increase QTL detection sensitivity. In parallel, I will measure molecular "fingerprints" in each segregant at several stages of aging using a technique pioneered by another EIPOD in the Steinmetz lab to isolate individuals of a specified age. These fingerprints will include measurements of the epigenome and regulatory DNA by DNase-seq, the transcriptome by RNA-sequencing, the proteome by SP3 MS developed at EMBL, and the metabolome by targeted LC-MS to quantify metabolites of central metabolism. These heterogeneous datasets will be combined and integrated with prior knowledge using a machine learning approach that I will build to identify combinations of molecular profiles that are diagnostic of the aging process in the context of particular genetic backgrounds. Finally, I will determine how the environment conditions genetic contributions to aging by leveraging a documented heat tolerance phenotype of YJM789. I will re-characterize QTLs and molecular fingerprints at elevated growth temperatures to assign an environmental (in)dependence to each of the genotypes and molecular reporters identified above.

Having established quantitative relationships between genetics, multi-omic molecular profiles, and the environment, it would be enticing to return full circle to provide actionable insights to human wellness companies. As a final component of my project, therefore, I would like to partner with a company called Arivale (a company started by the head of my previous Institute, Lee Hood) to adapt my computational approaches to discover molecular fingerprints of human wellness.

The Steinmetz group at EMBL is an ideal laboratory to host these experiments. The group has pioneered technologies to study yeast genetics. They maintain a fully genotyped collection of yeast segregants and have expertise in the sequencing approaches used in this project. The lab also has active collaborations to provide expertise for the proteomics and metabolomics experiments.

The Stegle group at EMBL-EBI will provide support for the computational components of this project. The Stegle lab has pioneered statistical analysis methods for genome wide association studies (GWAS). Their expertise will be critical to build a robust model that properly accounts for noise and biases unique to each data set.

I have expertise in systems biology, including both computational and experimental proficiencies. This project will constitute a second major transition in my career. With this project I will take on an interdisciplinary study at a much larger scale. Weaving together multiple layers of biological complexity will expand my computational abilities and allow me to learn new experimental skills, including microfluidics, metabolomics and proteomics. The EIPOD fellowship will be an indispensable, interdisciplinary support system that will make this project feasible.

1. Jo, MC, et al. 2015. PNAS. 112(3):9364-9369.
2. Stumpferl, SW, et al. 2012. Genome Res. 22(10):1963-73.
