* Please write a summary of your doctoral research and if applicable, of your postdoctoral scientific work. Indicate the significance, the aims and your contributions. (maximum 2000 characters)

During my PhD I studied gene regulatory networks (GRNs) in microorganisms. The structure of these networks describes which gene products regulate the production of which other genes. It is thought that the activity of these GRNs is responsible for generating expression patterns that help microorganisms cope with environmental change, e.g. ensuring that functionally related genes are expressed together. I developed algorithms to infer GRNs directly from high-throughput gene expression data for two phylogenetically-distant organisms (nearly 2,500 microarrays) with ensemble learning approaches. My models suggested that GRN function is highly environment-dependent. I was able to show how this environmental-dependence leads to extensive transcriptional heterogeneity across conditions. I reported that the patterns of which genes are co-expressed together across environments are flexible, i.e., a particular gene can be co-expressed independently with functionally different sets of genes depending on the environment. I demonstrated, for example, that nearly 1/4 of all E. coli operons have environment-dependent transcription isoforms, upending a long-standing dogma that operons are transcribed together as a single transcriptional unit. After developing my models, I built open-source software resources for researchers to construct and explore their own models. These efforts resulted in construction of a M. tuberculosis model 50X faster than the original models.

In April 2015 I joined Lars Steinmetz's group at EMBL to extend my interdisciplinary training, taking on a more complex model system (yeast) and learning how to measure cells at multiple molecular scales. Since joining the group I have been working to characterize yeast with multiple high-throughput technologies. This project is establishing some of the experimental workflows (transcriptomics, proteomics, and metabolomics) and first-generation computational pipelines that will be used for my proposed EIPOD project.
