Summary of previous research
(maximum of 1,700 characters including spaces and punctuation)

Summarize the research you carried out immediately prior to submitting this application. If you recently
finished your Ph.D., summarize your thesis work and describe any projects you will work on up to the
proposed activation date of the fellowship.

If you are an established postdoc, describe your current research.

If you are already working in the proposed host supervisor’s lab, describe the work you performed
in your previous laboratory and not the work you have done since you arrived in the proposed host lab.

This section should be written in non-technical language, and be understandable to scientists who are
not experts in your field.

NOTE – some information in this section is to be repeated in the “Achievements and research
goals” section. This is intentional; your career plans in the next section must be placed in the
context of previous training.

During my PhD I studied gene regulatory networks (GRNs) in microorganisms. The structure of these networks describes which gene products regulate the production of which other genes. It is thought that the activity of these GRNs is responsible for generating expression patterns that help microorganisms cope with environmental change, e.g. ensuring that functionally related genes are expressed together. I developed algorithms to infer GRNs directly from high-throughput gene expression data for two organisms (nearly 2,500 microarrays). My models suggested that GRN structure is highly environment-dependent. I was able to show how this environmental-dependence leads to extensive transcriptional heterogeneity across conditions. I reported that the patterns of which genes are co-expressed together across environments are flexible, i.e., a particular gene can be co-expressed independently with functionally different sets of genes depending on the environment. I demonstrated, for example, that nearly 1/4 of all E. coli operons have environment-dependent transcription isoforms, which contradicts a long-standing dogma that operons are transcribed together as a single transcriptional unit. This heterogeneity, it turns out, also has important phenotypic consequences. Genes that were co-discovered in modules we call condition-specific co-regulated modules (corems) often have very similar fitness consequences when they are deleted. After developing my models, I dedicated myself to building open-source software for researchers to construct and explore their own models. These efforts resulted in construction of a M. tuberculosis model 50X faster than the original models.
