Research plan, including references
(maximum of 15,000 characters including spaces and punctuation)

## Multiomic fingerprints of wellness in a genetically tractable yeast population

A new paradigm for personalized medicine is emerging in the biomedical sciences: quantified wellness. Rather than focus on treatment of disease by dissecting mechanisms and etiologies, this approach aims to monitor and quantify patient "wellness" in "real-time", suggesting personalized interventions to sustain health before a patient manifests disease [1]. This exciting new approach to medicine, however, raises a number of basic science questions that have yet to be addressed, primarily because the technologies that would enable such large-scale longitudinal monitoring of a biological system are newly developed, incompletely understood, and poorly integrated.  My proposal aims to develop a basic science understanding of how multiple omic-scale molecular readouts of cellular physiology should be combined and integrated with genetic information to establish molecular "fingerprints" predictive of large-scale phenotypes, such as health, using yeast as a tractable model system. Specifically, I will study a multifaceted process in yeast that approximates wellness, aging. The project will be conducted within the context of a tightly controlled genetic and experimental system so that I can directly evaluate predicted relationships predicted between genetics, longitudinal molecular fingerprints, the environment, and quantitative phenotypes using targeted genetic, molecular, and environmental interventions. I will combine multiple high-troughput molecular technologies with state-of-the-art computational and statistical methods to develop integrated and predictive models of cellular function.

The project is motivated by simultaneous changes in biology and technology that are reshaping the practice of medicine. For decades, biomedical research has focused on understanding and treating disease mechanisms and etiologies. This has been a particularly powerful way to develop drugs that directly interface with disease mechanisms. More recently, next-generation sequencing and related technologies have allowed investigators to achieve deeper insight into the genetics determinants of disease. Genome Wide Association Studies (GWAS), for example, have identified multiple genetic loci associated with several human diseases, ranging from disease with relatively simple pathophysiologies to much more complex scenarios. Despite many successes of these approaches, there are also notable shortcomings. In particular: (1) many diseases are polygenic with complex etiologies such that it is difficult to discern causal mechanism(s), (2) genetic loci identified by GWAS for complex diseases tend to have small effect sizes and oftentimes their precise relationship to disease is unclear, and (3) many common diseases may be driven by rare variations such that is difficult to identify the mechanism(s) responsible for disease in an individual. Unless addressed, these pitfalls will be magnified as the number and diversity of biological measurements increases. Apple, for example, has recently unveiled its ResearchKit, which will allow people using its mobile technologies to enroll in human research studies. This is just one of a growing number of technological developments that will dramatically change the scale of potential phenotyping []. There is, therefore, a growing need to understand how large collections of heterogeneous data can be integrated with genomic information to understand, model, and ultimately predict fundamental biological processes in individuals.

This is a question best addressed by basic science approaches. There are a number of fundamental scientific questions embedded within this problem that cut across disciplines, from mathematics to computer science and biochemistry - even ethical considerations become important. From a biological context, I will attempt to address several questions of particular interest to the field, including: (1) how the influence of genetics varies across molecular scales and (2) how genetic influences and their molecular fingerprints are shaped by the environment. I will conduct my experiments in a yeast population that is both genetically and experimentally tractable. My experiments will combine recent developments in omic-scale measurements with microfluidic systems that can image multiple features of single cells and precisely control their environment. Together, these data sets will allow me to quantifying relationships between genetic, molecular, and environmental factors that influence the aging process.

#### Experimental Design and Methodology

The experimental design for this project integrates both experimental methods from several disciplines with computation. I have an interdisciplinary background with expertise in both experimental and computational methods. Multiple high-throughput approaches will be used to quantify. Statistical learning approaches. Finally, targeted approaches will be used to validate predicted intervention points.

##### Yeast as a model system for aging

Yeast is a well established model of aging in eukaryotes. A yeastâ€™s longevity can be quantified as replicative life span (RLS), or the number of daughters produced before irreversible cell cycle arrest. High-throughput platforms have been established to measure RLS by isolating a single mother cell using microscopy and microfluidic devices [1]. As a first step in my project, I will establish an experimental system to measure RLS in a genetically diverse yeast population generated by the Steinmetz lab.  
##### Genetic determinants of aging in segregrating yeast population

Each of 140 haploid yeast segregants from a cross between the laboratory strain of S. cerevisiae, S288c, and a pathogenic strain isolated from the lungs of an AIDS patient, YJM789, will be characterized for RLS. A recent study identified SIR2 as a primary regulator of lifespan in yeast [2]. I will improve these measurements by using more sensitive QTL detection methods, a more finely measured phenotype, and segregants with greater genomic resolution to increase QTL detection sensitivity.

##### Multiomic fingerprints of wellness

###### Epigenomics

regulatory DNA by DNase-seq

###### Transcriptomics

To comprehensively profile yeast transcriptomes, we will leverage the well-established 5'PSeq and 3' Tagseq methods and corresponding computational analysis developed in the Steinmetz lab. The 3' Tagseq method allows the quantification and identification of polyadenylation sites, while 5'Pseq accurately determines the transcription start sites (27, 56). These methods provide robust estimations of transcript abundance levels and also reveal the complexity of the transcriptome. The additional information on transcript variation will help us to pinpoint mitochondrial phenotypes caused by transcript isoform variation.

###### Proteomeomics

Comprehensive proteomic data will be collected using data-dependent aquistion on a QExactive Orbitrap mass spectrometer. I will take advantage of the SP3 method optimized for proteomic analysis in yeast, to prepare cellular extracts for MS analysis (28). A spike-in standard will be included to normalize across all batches and strains. Subsequent steps including enzymatic digestion, LC-MS/MS measurement and computational analysis will be carried out using well-established protocols available in the Steinmetz lab.

###### Metabolomics

Intra- and extracellular metabolic data (absolute abundance of metabolites) is a multi-step process: we will reproducibly generate biomass, withdraw defined amounts of biological material, measure the absolute biovolume (total amount of cells surrounded by an intact membrane) (57) by volume displacement (coulter counter principle) (58)., precisely inactivate metabolism by cold methanol quenching (59), nonselectively extract (60) and measure the metabolites quantitatively. Targeted analysis of absolute intra- and extracellular metabolite concentrations will be performed using an LC-MS/MS. The measurement methods available on the QExactive Orbitrap LS-MS/MS cover almost all metabolites of the central carbon metabolism of yeast and provide absolute concentration via the use of isotopic labelled mass spectrometry coupled to chromatography.

###### Data integration and computational modeling

We will process the multi-omics data using established pipelines. After rigorous quality control, the data will be integrated into a single multi-omic dataset. To identify novel associations between specific genotypes and molecular phenotypes within this dataset, we will employ statistical learning techniques. Specifically, we will use ensemble learning methods, such as random forests, to discover meaningful associations across diverse molecular features (25). These predicted associations will be interpreted mechanistically within a systems-level model of metabolism and genetic regulation. Thus, we will be able to predict specific mechanisms through which several characterized genotypes lead to an observed phenotype. We will build this systems-level model by integrating up-to-date, public, genome-scale models of metabolism and genetic regulation in yeast using established methods (61).

In parallel, I will measure molecular "fingerprints" in each segregant at several stages of aging using a technique pioneered by another EIPOD in the Steinmetz lab to isolate individuals of a specified age. These fingerprints will include measurements of the epigenome and , the transcriptome by RNA-sequencing, the proteome by SP3 MS developed at EMBL, and the metabolome by targeted LC-MS to quantify metabolites of central metabolism. These heterogeneous datasets will be combined and integrated with prior knowledge using a machine learning approach that I will build to identify combinations of molecular profiles that are diagnostic of the aging process in the context of particular genetic backgrounds. Finally, I will determine how the environment conditions genetic contributions to aging by leveraging a documented heat tolerance phenotype of YJM789. I will re-characterize QTLs and molecular fingerprints at elevated growth temperatures to assign an environmental (in)dependence to each of the genotypes and molecular reporters identified above.

Having established quantitative relationships between genetics, multi-omic molecular profiles, and the environment, it would be enticing to return full circle to provide actionable insights to human wellness companies. As a final component of my project, therefore, I would like to partner with a company called Arivale (a company started by the head of my previous Institute, Lee Hood) to adapt my computational approaches to discover molecular fingerprints of human wellness.

##### Environmental dependence

##### Validation

#### Risk management and feasibility

#### Significance

The project generated will open source framework that can be used to study, including human wellness. The computational component will consist of statistical learning approaches that can be generalized to other types of data, such as those routinely collected from human patients. Second, and more importantly, I will use the data and integrated models to address two fundamental biological questions that are of particular interest to me. Of particular intest to me are two fundamntal:  

[1] Hood L, Price N. Demystifying Disease, Democratizing Healthcare. (2014) Science Translational Medicine. 6:225
[2] The coming era of human phenotyping. (2015) Nat Biotechnol. 33(6):567.
