# EMBL Project Summary
###### Aaron Brooks
###### Advisor: Lars Steinmetz, EMBL Heidelberg
###### 15.07.2015

## Deciphering Causal Pathways That Mediate Genetic Effects on Multiomic Molecular Profiles 

*with Chenchen Zhu (EMBL), Nicole Paczia (ICSB), Carole Linster (ICSB), and Oliver Stegle (EMBL-EBI)* 

Unraveling molecular processes that lead from genotype to phenotype is crucial to treat genetic diseases as well as understand how molecular pathways evolve. Genetic determinants of phenotypes and disease are typically mapped as quantitative trait loci (QTLs) in Genome Wide Association Studies (GWAS). Until recently few studies have considered the effect of genetic variation at more than one biological scale. Furthermore, to our knowledge none of these studies has been able to place a QTL within an integrated cellular context to determine how it functions.  

*My project will identify genetic variations that affect molecular processes across multiple biological layers, from transcription to post-translational mechanisms. I will place the variants we identify within the context of integrated biochemical networks to reveal intervention points that can modulate their effect.*

So far we have profiled the transcriptome, proteome, and metabolome for 50 genetically diverse yeast strains. We measured variation in transcript levels, 3Â´ transcript isoform usage, protein levels, and the dynamics of more than 30 metabolites at four time points during exponential growth. Our dataset provides an unprecedented window into the consequences of genetic variation across several molecular layers. My task will be to integrate these data using machine learning techniques, leveraging prior knowledge to robustly detect multi-omic QTLs and interpret their function within the context of macromolecular networks.  

The findings from this study will have implications for the design and analysis of clinical omics studies aimed at discovering personalized targets for molecular intervention. We anticipate that combining measurements across multiple molecular profiles and integrating these measurements within the context of biological networks will increase our ability to detect QTLs and understand how they function.  

Goals: 
- Develop bioinformatic methods to identify multiomic QTLs from heterogeneous molecular profiles (i.e., transcriptomics, proteomics, and metabolomics)
- Identify QTLs that affect individual processes (e.g., transcription), as well as those that have an affect across multiple molecular layers (e.g., from transcript to protein levels)
- Place identified QTLs within the context of macromolecular networks to implicate pathways that can be targeted to modify the effect of a QTL 
- Validate several intervention points using targeted allele replacement/deletion (e.g. CRISPR-Cas9)