{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ARMseq Pipeline\n",
    "\n",
    "Getting ARMseq pipeline to work. Yeast data.\n",
    "\n",
    "Note: not all commands can be excuted from this script. See Preliminaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    ".libPaths(\"~/R/x86_64-redhat-linux-gnu-library/3.2/\")\n",
    "options(repr.plot.width = 6)\n",
    "options(repr.plot.height = 5)\n",
    "options(jupyter.plot_mimetypes = 'image/png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,\n",
      "    do.call, duplicated, eval, evalq, Filter, Find, get, intersect,\n",
      "    is.unsorted, lapply, Map, mapply, match, mget, order, paste, pmax,\n",
      "    pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rep.int,\n",
      "    rownames, sapply, setdiff, sort, table, tapply, union, unique,\n",
      "    unlist, unsplit\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Creating a generic function for ‘nchar’ from package ‘base’ in package ‘S4Vectors’\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    rename\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: XVector\n",
      "Loading required package: Biostrings\n",
      "Loading required package: ade4\n",
      "\n",
      "Attaching package: ‘ade4’\n",
      "\n",
      "The following object is masked from ‘package:Biostrings’:\n",
      "\n",
      "    score\n",
      "\n",
      "The following object is masked from ‘package:rtracklayer’:\n",
      "\n",
      "    score\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    score\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    score\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    score\n",
      "\n",
      "\n",
      "Attaching package: ‘seqinr’\n",
      "\n",
      "The following object is masked from ‘package:Biostrings’:\n",
      "\n",
      "    translate\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    count, query\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(rtracklayer)\n",
    "library(ggplot2)\n",
    "library(Rsamtools)\n",
    "library(Biostrings)\n",
    "library(seqinr)\n",
    "library(parallel)\n",
    "options(mc.cores = 24)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SHA-1 hash of file is 7565259173739547b4f1cbdfab1ca6430db08508\n"
     ]
    }
   ],
   "source": [
    "devtools::source_url(\"https://raw.githubusercontent.com/scalefreegan/steinmetz-lab/master/yeast2_0/scripts/seqtools.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "genome = read.fasta(\"/g/steinmetz/genome/Saccharomyces_cerevisiae/S288c_ERCC/ver_20150412/fasta/S288c_complete_wERCC.fsa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Preliminaries\n",
    "\n",
    "First you must download ARMseq datasets with SRA toolkit (`fastq-dump`), cut adapter sequences with `cutadapt`, and then align, sort, and index the reads using `bowtie2`. [This](processANDalign.txt) bash script does these steps. Final output are bam files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remapping in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# working directory\n",
    "setwd(\"/g/steinmetz/project/IESY/trna/ARMseq/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load tRNA annotations from [GtRNAdb](http://gtrnadb.ucsc.edu/genomes/eukaryota/Scere3/)\n",
    "\n",
    "*Extend upstream and downstream regions with `window` parameter*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "window = 50\n",
    "\n",
    "tRNA_annotations = read.delim(\n",
    "    \"../ARMseq/data/tRNA_isoforms_wleaders_4R.txt\", sep = \"\\t\", header = T, strip.white=TRUE)\n",
    "# switch start / end when end < start. denote strand instead\n",
    "tRNA_annotations = data.frame(do.call(rbind, apply(tRNA_annotations, 1, function(i){\n",
    "    #print(i[\"Start\"])\n",
    "    if (i[\"Chr\"] == \"chrM\") {\n",
    "        i[\"Chr\"] = \"chrMT\"\n",
    "    }\n",
    "    if (as.numeric(i[\"Start\"]) < as.numeric(i[\"End\"])) {\n",
    "        o = i\n",
    "        # modify upstream/downstream sequences based on window of interest\n",
    "        o$Upstream = paste(genome[[i[\"Chr\"]]][seq((as.numeric(i[\"Start\"]) - window),\n",
    "                                                          (as.numeric(i[\"Start\"]) - 1))], collapse = \"\")\n",
    "        o$Downstream = paste(genome[[i[\"Chr\"]]][seq((as.numeric(i[\"End\"]) + 1),\n",
    "                                                          (as.numeric(i[\"End\"]) + window))], collapse = \"\")\n",
    "        o$Start = as.numeric(i[\"Start\"])\n",
    "        o$End = as.numeric(i[\"End\"])\n",
    "        o$Strand = \"+\"\n",
    "    } else {\n",
    "        o = i\n",
    "        # modify upstream/downstream sequences based on window of interest\n",
    "        o$Upstream = tolower(as.character(reverseComplement(DNAString(paste(genome[[i[\"Chr\"]]][seq((as.numeric(i[\"Start\"]) + 1),\n",
    "                                                          (as.numeric(i[\"Start\"]) + window))], collapse = \"\")))))\n",
    "        o$Downstream = tolower(as.character(reverseComplement(DNAString(paste(genome[[i[\"Chr\"]]][seq((as.numeric(i[\"End\"]) - window),\n",
    "                                                          (as.numeric(i[\"End\"]) - 1))], collapse = \"\")))))\n",
    "        o$Start = as.numeric(i[\"End\"])\n",
    "        o$End = as.numeric(i[\"Start\"])\n",
    "        o$Strand = \"-\"\n",
    "    }\n",
    "    return(o)\n",
    "})))\n",
    "tRNA_annotations$Start = as.numeric(tRNA_annotations$Start)\n",
    "tRNA_annotations$End = as.numeric(tRNA_annotations$End)\n",
    "tRNA_annotations$End = as.numeric(tRNA_annotations$End)\n",
    "tRNA_annotations$tRNA.Number = as.numeric(tRNA_annotations$tRNA.Number)\n",
    "tRNA_annotations$tRNA.type = as.character(tRNA_annotations$tRNA.type)\n",
    "tRNA_annotations$Anticodon = unlist(tRNA_annotations$Anticodon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>Chr</th><th scope=col>tRNA.Number</th><th scope=col>Start</th><th scope=col>End</th><th scope=col>tRNA.type</th><th scope=col>Anticodon</th><th scope=col>Upstream</th><th scope=col>Downstream</th><th scope=col>Intron.Start</th><th scope=col>Intron.End</th><th scope=col>Cove</th><th scope=col>HMM</th><th scope=col>X2pStr</th><th scope=col>Hit.Origin</th><th scope=col>Strand</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chrIV</td><td>3</td><td>410379</td><td>410451</td><td>Ala</td><td>AGC</td><td>ttaagtgcaaacatgtaagaaagattctcggtaacgaccatacaaatatt</td><td>aattattttttactttccgcggtgccgagatgcagacgtggccaactgtg</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chrVI</td><td>6</td><td>204924</td><td>204996</td><td>Ala</td><td>AGC</td><td>tgtaaagtaaacctgtatcatttctgtttggaagatttcaacaaacgttc</td><td>tttaaatttttttattttttttgtcaaatctctctatctatagagataat</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chrVII</td><td>15</td><td>774349</td><td>774421</td><td>Ala</td><td>AGC</td><td>ttagatttacttcattattcttctctattctcccctttcatcaattaatt</td><td>aatttttttacttttaatgaggtatactggaatattacggctcttttttt</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chrVIII</td><td>8</td><td>146242</td><td>146314</td><td>Ala</td><td>AGC</td><td>acactccggtattactcgagcccgtaatacaacacattcaacaaatatca</td><td>atcttttttgttattaccataacagtactataaaggtctgttaatgtata</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chrX</td><td>22</td><td>197313</td><td>197385</td><td>Ala</td><td>AGC</td><td>tgtcatactgacgtatctcactttgataaacaacaatatcaataatatta</td><td>atttctttttgttctcaacgacaataataagaaaatcaacttaaatggaa</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chrXI</td><td>4</td><td>219895</td><td>219967</td><td>Ala</td><td>AGC</td><td>taaaagagcttgttatagtaatctaagtggaaatacactaacagtaaata</td><td>atctttttatacttattaataatttttttcctgccgttacttgcttttaa</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>+</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & Chr & tRNA.Number & Start & End & tRNA.type & Anticodon & Upstream & Downstream & Intron.Start & Intron.End & Cove & HMM & X2pStr & Hit.Origin & Strand\\\\\n",
       "\\hline\n",
       "\t1 & chrIV & 3 & 410379 & 410451 & Ala & AGC & ttaagtgcaaacatgtaagaaagattctcggtaacgaccatacaaatatt & aattattttttactttccgcggtgccgagatgcagacgtggccaactgtg &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & +\\\\\n",
       "\t2 & chrVI & 6 & 204924 & 204996 & Ala & AGC & tgtaaagtaaacctgtatcatttctgtttggaagatttcaacaaacgttc & tttaaatttttttattttttttgtcaaatctctctatctatagagataat &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & -\\\\\n",
       "\t3 & chrVII & 15 & 774349 & 774421 & Ala & AGC & ttagatttacttcattattcttctctattctcccctttcatcaattaatt & aatttttttacttttaatgaggtatactggaatattacggctcttttttt &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & +\\\\\n",
       "\t4 & chrVIII & 8 & 146242 & 146314 & Ala & AGC & acactccggtattactcgagcccgtaatacaacacattcaacaaatatca & atcttttttgttattaccataacagtactataaaggtctgttaatgtata &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & -\\\\\n",
       "\t5 & chrX & 22 & 197313 & 197385 & Ala & AGC & tgtcatactgacgtatctcactttgataaacaacaatatcaataatatta & atttctttttgttctcaacgacaataataagaaaatcaacttaaatggaa &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & -\\\\\n",
       "\t6 & chrXI & 4 & 219895 & 219967 & Ala & AGC & taaaagagcttgttatagtaatctaagtggaaatacactaacagtaaata & atctttttatacttattaataatttttttcctgccgttacttgcttttaa &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & +\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "      Chr tRNA.Number  Start    End tRNA.type Anticodon\n",
       "1   chrIV           3 410379 410451       Ala       AGC\n",
       "2   chrVI           6 204924 204996       Ala       AGC\n",
       "3  chrVII          15 774349 774421       Ala       AGC\n",
       "4 chrVIII           8 146242 146314       Ala       AGC\n",
       "5    chrX          22 197313 197385       Ala       AGC\n",
       "6   chrXI           4 219895 219967       Ala       AGC\n",
       "                                            Upstream\n",
       "1 ttaagtgcaaacatgtaagaaagattctcggtaacgaccatacaaatatt\n",
       "2 tgtaaagtaaacctgtatcatttctgtttggaagatttcaacaaacgttc\n",
       "3 ttagatttacttcattattcttctctattctcccctttcatcaattaatt\n",
       "4 acactccggtattactcgagcccgtaatacaacacattcaacaaatatca\n",
       "5 tgtcatactgacgtatctcactttgataaacaacaatatcaataatatta\n",
       "6 taaaagagcttgttatagtaatctaagtggaaatacactaacagtaaata\n",
       "                                          Downstream Intron.Start Intron.End\n",
       "1 aattattttttactttccgcggtgccgagatgcagacgtggccaactgtg            0          0\n",
       "2 tttaaatttttttattttttttgtcaaatctctctatctatagagataat            0          0\n",
       "3 aatttttttacttttaatgaggtatactggaatattacggctcttttttt            0          0\n",
       "4 atcttttttgttattaccataacagtactataaaggtctgttaatgtata            0          0\n",
       "5 atttctttttgttctcaacgacaataataagaaaatcaacttaaatggaa            0          0\n",
       "6 atctttttatacttattaataatttttttcctgccgttacttgcttttaa            0          0\n",
       "   Cove   HMM X2pStr Hit.Origin Strand\n",
       "1 74.48 50.69  23.79         Bo      +\n",
       "2 74.48 50.69  23.79         Bo      -\n",
       "3 74.48 50.69  23.79         Bo      +\n",
       "4 74.48 50.69  23.79         Bo      -\n",
       "5 74.48 50.69  23.79         Bo      -\n",
       "6 74.48 50.69  23.79         Bo      +"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "head(tRNA_annotations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract only reads aligned to these regions from bam files\n",
    "\n",
    "*Steps*\n",
    "1. Make GRanges object\n",
    "2. Convert to RangesList\n",
    "3. Use `scanBam` from `Rsamtools` to extract alignments\n",
    "4. Process bam file so that it can be combined with tRNA table above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "window2 = 50\n",
    "\n",
    "gr <- GRanges(\n",
    "    seqnames = Rle(unlist(tRNA_annotations$Chr)),\n",
    "    ranges = IRanges(start = tRNA_annotations$Start - window2,\n",
    "                 end = tRNA_annotations$End + window2),\n",
    "    strand = Rle(unlist(tRNA_annotations$Strand)),\n",
    "    mcols = select(tRNA_annotations, tRNA.Number,tRNA.type, Upstream, Downstream)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gr_list = RangesList(lapply(seq(1,length(gr)),function(i){ranges(gr[i,])}))\n",
    "names(gr_list) = seqnames(gr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import bam files for the following samples/alignments:\n",
    "\n",
    "- AlkB treated\n",
    "- AlkB trated with very-sensitive, end-to-end alignment\n",
    "- Untreated\n",
    "- Untreated with very-sensitive, end-to-end alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In `levels<-`(`*tmp*`, value = if (nl == nL) as.character(labels) else paste0(labels, : duplicated levels in factors are deprecatedWarning message:\n",
      "In `levels<-`(`*tmp*`, value = if (nl == nL) as.character(labels) else paste0(labels, : duplicated levels in factors are deprecatedWarning message:\n",
      "In `levels<-`(`*tmp*`, value = if (nl == nL) as.character(labels) else paste0(labels, : duplicated levels in factors are deprecatedWarning message:\n",
      "In `levels<-`(`*tmp*`, value = if (nl == nL) as.character(labels) else paste0(labels, : duplicated levels in factors are deprecated"
     ]
    }
   ],
   "source": [
    "fdir = \"/g/steinmetz/project/IESY/trna/ARMseq/data/\" # where are the bam files\n",
    "what = c(\"qname\", \"rname\", \"strand\", \"pos\", \"mapq\", \"cigar\", \"seq\", \"qual\") # what to access from alignments\n",
    "tags = c(\"AS\", \"XS\", \"NM\")\n",
    "param = ScanBamParam(which = gr_list, what = what, tag = tags)\n",
    "\n",
    "\n",
    "untreated <- scanBam(paste0(fdir,\"untreated.bam\"), param=param, \n",
    "               index = paste0(fdir,\"untreated.bam.bai\"))\n",
    "#untreated_vse2e <- scanBam(paste0(fdir,\"untreated_vse2e.bam\"), param=param, \n",
    "#               index = paste0(fdir,\"untreated_vse2e.bam.bai\"))\n",
    "# alkB <- scanBam(paste0(fdir,\"treated.bam\"), param=param, \n",
    "#                index = paste0(fdir,\"treated.bam.bai\"))\n",
    "# alkB_vse2e <- scanBam(paste0(fdir,\"treated_vse2e.bam\"), param=param, \n",
    "#                index = paste0(fdir,\"treated_vse2e.bam.bai\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alter bam data.frame to make it compatible with tRNA data.frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "untreated = makeHappyBam(untreated, tRNA_annotations)\n",
    "#untreated_vse2e = makeHappyBam(untreated_vse2e, tRNA_annotations)\n",
    "#alkB = makeHappyBam(alkB)\n",
    "#alkB_vse2e = makeHappyBam(alkB_vse2e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Try to assign counts for all reads (tRNAs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>Chr</th><th scope=col>Start</th><th scope=col>End</th><th scope=col>tRNA.Number</th><th scope=col>tRNA.type</th><th scope=col>Anticodon</th><th scope=col>Upstream</th><th scope=col>Downstream</th><th scope=col>Intron.Start</th><th scope=col>Intron.End</th><th scope=col>Cove</th><th scope=col>HMM</th><th scope=col>X2pStr</th><th scope=col>Hit.Origin</th><th scope=col>Strand</th><th scope=col>qname</th><th scope=col>rname</th><th scope=col>strand</th><th scope=col>pos</th><th scope=col>mapq</th><th scope=col>cigar</th><th scope=col>seq</th><th scope=col>qual</th><th scope=col>AS</th><th scope=col>XS</th><th scope=col>NM</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chrIV</td><td>410379</td><td>410451</td><td>3</td><td>Ala</td><td>AGC</td><td>ttaagtgcaaacatgtaagaaagattctcggtaacgaccatacaaatatt</td><td>aattattttttactttccgcggtgccgagatgcagacgtggccaactgtg</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>+</td><td>SRR1874045.292256</td><td>chrIV</td><td>+</td><td>410430</td><td>0</td><td>25M</td><td>GGTTCGATTCCGGACTCGTCCACTA</td><td>BCABBCCFFFFFGGGGGGGGGEHHH</td><td>-15</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chrVI</td><td>204924</td><td>204996</td><td>6</td><td>Ala</td><td>AGC</td><td>tgtaaagtaaacctgtatcatttctgtttggaagatttcaacaaacgttc</td><td>tttaaatttttttattttttttgtcaaatctctctatctatagagataat</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td><td>SRR1874045.292256</td><td>chrVI</td><td>-</td><td>204921</td><td>0</td><td>25M</td><td>TAGTGGACGAGTCCGGAATCGAACC</td><td>HHHEGGGGGGGGGFFFFFCCBBACB</td><td>-10</td><td>NA</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chrVII</td><td>774349</td><td>774421</td><td>15</td><td>Ala</td><td>AGC</td><td>ttagatttacttcattattcttctctattctcccctttcatcaattaatt</td><td>aatttttttacttttaatgaggtatactggaatattacggctcttttttt</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>+</td><td>SRR1874045.292256</td><td>chrVII</td><td>+</td><td>774400</td><td>0</td><td>25M</td><td>GGTTCGATTCCGGACTCGTCCACTA</td><td>BCABBCCFFFFFGGGGGGGGGEHHH</td><td>-15</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chrVIII</td><td>146242</td><td>146314</td><td>8</td><td>Ala</td><td>AGC</td><td>acactccggtattactcgagcccgtaatacaacacattcaacaaatatca</td><td>atcttttttgttattaccataacagtactataaaggtctgttaatgtata</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td><td>SRR1874045.292256</td><td>chrVIII</td><td>-</td><td>146239</td><td>0</td><td>25M</td><td>TAGTGGACGAGTCCGGAATCGAACC</td><td>HHHEGGGGGGGGGFFFFFCCBBACB</td><td>-10</td><td>NA</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chrX</td><td>197313</td><td>197385</td><td>22</td><td>Ala</td><td>AGC</td><td>tgtcatactgacgtatctcactttgataaacaacaatatcaataatatta</td><td>atttctttttgttctcaacgacaataataagaaaatcaacttaaatggaa</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td><td>SRR1874045.292256</td><td>chrX</td><td>-</td><td>197310</td><td>0</td><td>25M</td><td>TAGTGGACGAGTCCGGAATCGAACC</td><td>HHHEGGGGGGGGGFFFFFCCBBACB</td><td>-10</td><td>NA</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chrXI</td><td>219895</td><td>219967</td><td>4</td><td>Ala</td><td>AGC</td><td>taaaagagcttgttatagtaatctaagtggaaatacactaacagtaaata</td><td>atctttttatacttattaataatttttttcctgccgttacttgcttttaa</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>+</td><td>SRR1874045.292256</td><td>chrXI</td><td>+</td><td>219946</td><td>0</td><td>25M</td><td>GGTTCGATTCCGGACTCGTCCACTA</td><td>BCABBCCFFFFFGGGGGGGGGEHHH</td><td>-10</td><td>NA</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>chrXI</td><td>517988</td><td>518060</td><td>9</td><td>Ala</td><td>AGC</td><td>gtaaaaatatatgatccaattcctaggacatcagaaaatatatctactat</td><td>ttttaaaatttctttttttttttttctctttccgtatgtaagcttttttc</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>+</td><td>SRR1874045.292256</td><td>chrXI</td><td>+</td><td>518039</td><td>0</td><td>25M</td><td>GGTTCGATTCCGGACTCGTCCACTA</td><td>BCABBCCFFFFFGGGGGGGGGEHHH</td><td>-10</td><td>NA</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>chrXII</td><td>656934</td><td>657006</td><td>16</td><td>Ala</td><td>AGC</td><td>tatcatactgacgtaccccattctgaaacacaacagacaacaagaaaatt</td><td>agtaaatattttttgcaacttcttcaacctccaattattattcgatcatt</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td><td>SRR1874045.292256</td><td>chrXII</td><td>-</td><td>656931</td><td>0</td><td>25M</td><td>TAGTGGACGAGTCCGGAATCGAACC</td><td>HHHEGGGGGGGGGFFFFFCCBBACB</td><td>-15</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>chrXIII</td><td>321147</td><td>321219</td><td>17</td><td>Ala</td><td>AGC</td><td>tatcctacatttatgttttctatacttcttcaaagtcttcaatacaaatc</td><td>ttttttatttttttggctaactaacccatatataacgtcttttacgaggt</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td><td>SRR1874045.292256</td><td>chrXIII</td><td>-</td><td>321144</td><td>0</td><td>25M</td><td>TAGTGGACGAGTCCGGAATCGAACC</td><td>HHHEGGGGGGGGGFFFFFCCBBACB</td><td>-10</td><td>NA</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>chrXIII</td><td>768369</td><td>768441</td><td>9</td><td>Ala</td><td>AGC</td><td>ttttaaactgatttcactttttacagtggctagaaattacaataagaaat</td><td>aattttttgctactttgcaattagtgtaacttcatatttaaaattatagt</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>-</td><td>SRR1874045.292256</td><td>chrXIII</td><td>-</td><td>768366</td><td>0</td><td>25M</td><td>TAGTGGACGAGTCCGGAATCGAACC</td><td>HHHEGGGGGGGGGFFFFFCCBBACB</td><td>-15</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>chrXVI</td><td>856902</td><td>856974</td><td>7</td><td>Ala</td><td>AGC</td><td>acactccggtattactcgagcccgtaatacaacaatttacaacaagattc</td><td>ttttaatttttgattatggaatacaaaatgtatatttcgtataaaattat</td><td>      0</td><td>      0</td><td>74.48</td><td>50.69</td><td>23.79</td><td>Bo</td><td>+</td><td>SRR1874045.292256</td><td>chrXVI</td><td>+</td><td>856953</td><td>0</td><td>25M</td><td>GGTTCGATTCCGGACTCGTCCACTA</td><td>BCABBCCFFFFFGGGGGGGGGEHHH</td><td>-10</td><td>NA</td><td>2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllllllllllllll}\n",
       "  & Chr & Start & End & tRNA.Number & tRNA.type & Anticodon & Upstream & Downstream & Intron.Start & Intron.End & Cove & HMM & X2pStr & Hit.Origin & Strand & qname & rname & strand & pos & mapq & cigar & seq & qual & AS & XS & NM\\\\\n",
       "\\hline\n",
       "\t1 & chrIV & 410379 & 410451 & 3 & Ala & AGC & ttaagtgcaaacatgtaagaaagattctcggtaacgaccatacaaatatt & aattattttttactttccgcggtgccgagatgcagacgtggccaactgtg &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & + & SRR1874045.292256 & chrIV & + & 410430 & 0 & 25M & GGTTCGATTCCGGACTCGTCCACTA & BCABBCCFFFFFGGGGGGGGGEHHH & -15 & NA & 3\\\\\n",
       "\t2 & chrVI & 204924 & 204996 & 6 & Ala & AGC & tgtaaagtaaacctgtatcatttctgtttggaagatttcaacaaacgttc & tttaaatttttttattttttttgtcaaatctctctatctatagagataat &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & - & SRR1874045.292256 & chrVI & - & 204921 & 0 & 25M & TAGTGGACGAGTCCGGAATCGAACC & HHHEGGGGGGGGGFFFFFCCBBACB & -10 & NA & 2\\\\\n",
       "\t3 & chrVII & 774349 & 774421 & 15 & Ala & AGC & ttagatttacttcattattcttctctattctcccctttcatcaattaatt & aatttttttacttttaatgaggtatactggaatattacggctcttttttt &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & + & SRR1874045.292256 & chrVII & + & 774400 & 0 & 25M & GGTTCGATTCCGGACTCGTCCACTA & BCABBCCFFFFFGGGGGGGGGEHHH & -15 & NA & 3\\\\\n",
       "\t4 & chrVIII & 146242 & 146314 & 8 & Ala & AGC & acactccggtattactcgagcccgtaatacaacacattcaacaaatatca & atcttttttgttattaccataacagtactataaaggtctgttaatgtata &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & - & SRR1874045.292256 & chrVIII & - & 146239 & 0 & 25M & TAGTGGACGAGTCCGGAATCGAACC & HHHEGGGGGGGGGFFFFFCCBBACB & -10 & NA & 2\\\\\n",
       "\t5 & chrX & 197313 & 197385 & 22 & Ala & AGC & tgtcatactgacgtatctcactttgataaacaacaatatcaataatatta & atttctttttgttctcaacgacaataataagaaaatcaacttaaatggaa &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & - & SRR1874045.292256 & chrX & - & 197310 & 0 & 25M & TAGTGGACGAGTCCGGAATCGAACC & HHHEGGGGGGGGGFFFFFCCBBACB & -10 & NA & 2\\\\\n",
       "\t6 & chrXI & 219895 & 219967 & 4 & Ala & AGC & taaaagagcttgttatagtaatctaagtggaaatacactaacagtaaata & atctttttatacttattaataatttttttcctgccgttacttgcttttaa &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & + & SRR1874045.292256 & chrXI & + & 219946 & 0 & 25M & GGTTCGATTCCGGACTCGTCCACTA & BCABBCCFFFFFGGGGGGGGGEHHH & -10 & NA & 2\\\\\n",
       "\t7 & chrXI & 517988 & 518060 & 9 & Ala & AGC & gtaaaaatatatgatccaattcctaggacatcagaaaatatatctactat & ttttaaaatttctttttttttttttctctttccgtatgtaagcttttttc &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & + & SRR1874045.292256 & chrXI & + & 518039 & 0 & 25M & GGTTCGATTCCGGACTCGTCCACTA & BCABBCCFFFFFGGGGGGGGGEHHH & -10 & NA & 2\\\\\n",
       "\t8 & chrXII & 656934 & 657006 & 16 & Ala & AGC & tatcatactgacgtaccccattctgaaacacaacagacaacaagaaaatt & agtaaatattttttgcaacttcttcaacctccaattattattcgatcatt &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & - & SRR1874045.292256 & chrXII & - & 656931 & 0 & 25M & TAGTGGACGAGTCCGGAATCGAACC & HHHEGGGGGGGGGFFFFFCCBBACB & -15 & NA & 3\\\\\n",
       "\t9 & chrXIII & 321147 & 321219 & 17 & Ala & AGC & tatcctacatttatgttttctatacttcttcaaagtcttcaatacaaatc & ttttttatttttttggctaactaacccatatataacgtcttttacgaggt &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & - & SRR1874045.292256 & chrXIII & - & 321144 & 0 & 25M & TAGTGGACGAGTCCGGAATCGAACC & HHHEGGGGGGGGGFFFFFCCBBACB & -10 & NA & 2\\\\\n",
       "\t10 & chrXIII & 768369 & 768441 & 9 & Ala & AGC & ttttaaactgatttcactttttacagtggctagaaattacaataagaaat & aattttttgctactttgcaattagtgtaacttcatatttaaaattatagt &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & - & SRR1874045.292256 & chrXIII & - & 768366 & 0 & 25M & TAGTGGACGAGTCCGGAATCGAACC & HHHEGGGGGGGGGFFFFFCCBBACB & -15 & NA & 3\\\\\n",
       "\t11 & chrXVI & 856902 & 856974 & 7 & Ala & AGC & acactccggtattactcgagcccgtaatacaacaatttacaacaagattc & ttttaatttttgattatggaatacaaaatgtatatttcgtataaaattat &       0 &       0 & 74.48 & 50.69 & 23.79 & Bo & + & SRR1874045.292256 & chrXVI & + & 856953 & 0 & 25M & GGTTCGATTCCGGACTCGTCCACTA & BCABBCCFFFFFGGGGGGGGGEHHH & -10 & NA & 2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "       Chr  Start    End tRNA.Number tRNA.type Anticodon\n",
       "1    chrIV 410379 410451           3       Ala       AGC\n",
       "2    chrVI 204924 204996           6       Ala       AGC\n",
       "3   chrVII 774349 774421          15       Ala       AGC\n",
       "4  chrVIII 146242 146314           8       Ala       AGC\n",
       "5     chrX 197313 197385          22       Ala       AGC\n",
       "6    chrXI 219895 219967           4       Ala       AGC\n",
       "7    chrXI 517988 518060           9       Ala       AGC\n",
       "8   chrXII 656934 657006          16       Ala       AGC\n",
       "9  chrXIII 321147 321219          17       Ala       AGC\n",
       "10 chrXIII 768369 768441           9       Ala       AGC\n",
       "11  chrXVI 856902 856974           7       Ala       AGC\n",
       "                                             Upstream\n",
       "1  ttaagtgcaaacatgtaagaaagattctcggtaacgaccatacaaatatt\n",
       "2  tgtaaagtaaacctgtatcatttctgtttggaagatttcaacaaacgttc\n",
       "3  ttagatttacttcattattcttctctattctcccctttcatcaattaatt\n",
       "4  acactccggtattactcgagcccgtaatacaacacattcaacaaatatca\n",
       "5  tgtcatactgacgtatctcactttgataaacaacaatatcaataatatta\n",
       "6  taaaagagcttgttatagtaatctaagtggaaatacactaacagtaaata\n",
       "7  gtaaaaatatatgatccaattcctaggacatcagaaaatatatctactat\n",
       "8  tatcatactgacgtaccccattctgaaacacaacagacaacaagaaaatt\n",
       "9  tatcctacatttatgttttctatacttcttcaaagtcttcaatacaaatc\n",
       "10 ttttaaactgatttcactttttacagtggctagaaattacaataagaaat\n",
       "11 acactccggtattactcgagcccgtaatacaacaatttacaacaagattc\n",
       "                                           Downstream Intron.Start Intron.End\n",
       "1  aattattttttactttccgcggtgccgagatgcagacgtggccaactgtg            0          0\n",
       "2  tttaaatttttttattttttttgtcaaatctctctatctatagagataat            0          0\n",
       "3  aatttttttacttttaatgaggtatactggaatattacggctcttttttt            0          0\n",
       "4  atcttttttgttattaccataacagtactataaaggtctgttaatgtata            0          0\n",
       "5  atttctttttgttctcaacgacaataataagaaaatcaacttaaatggaa            0          0\n",
       "6  atctttttatacttattaataatttttttcctgccgttacttgcttttaa            0          0\n",
       "7  ttttaaaatttctttttttttttttctctttccgtatgtaagcttttttc            0          0\n",
       "8  agtaaatattttttgcaacttcttcaacctccaattattattcgatcatt            0          0\n",
       "9  ttttttatttttttggctaactaacccatatataacgtcttttacgaggt            0          0\n",
       "10 aattttttgctactttgcaattagtgtaacttcatatttaaaattatagt            0          0\n",
       "11 ttttaatttttgattatggaatacaaaatgtatatttcgtataaaattat            0          0\n",
       "    Cove   HMM X2pStr Hit.Origin Strand             qname   rname strand    pos\n",
       "1  74.48 50.69  23.79         Bo      + SRR1874045.292256   chrIV      + 410430\n",
       "2  74.48 50.69  23.79         Bo      - SRR1874045.292256   chrVI      - 204921\n",
       "3  74.48 50.69  23.79         Bo      + SRR1874045.292256  chrVII      + 774400\n",
       "4  74.48 50.69  23.79         Bo      - SRR1874045.292256 chrVIII      - 146239\n",
       "5  74.48 50.69  23.79         Bo      - SRR1874045.292256    chrX      - 197310\n",
       "6  74.48 50.69  23.79         Bo      + SRR1874045.292256   chrXI      + 219946\n",
       "7  74.48 50.69  23.79         Bo      + SRR1874045.292256   chrXI      + 518039\n",
       "8  74.48 50.69  23.79         Bo      - SRR1874045.292256  chrXII      - 656931\n",
       "9  74.48 50.69  23.79         Bo      - SRR1874045.292256 chrXIII      - 321144\n",
       "10 74.48 50.69  23.79         Bo      - SRR1874045.292256 chrXIII      - 768366\n",
       "11 74.48 50.69  23.79         Bo      + SRR1874045.292256  chrXVI      + 856953\n",
       "   mapq cigar                       seq                      qual  AS XS NM\n",
       "1     0   25M GGTTCGATTCCGGACTCGTCCACTA BCABBCCFFFFFGGGGGGGGGEHHH -15 NA  3\n",
       "2     0   25M TAGTGGACGAGTCCGGAATCGAACC HHHEGGGGGGGGGFFFFFCCBBACB -10 NA  2\n",
       "3     0   25M GGTTCGATTCCGGACTCGTCCACTA BCABBCCFFFFFGGGGGGGGGEHHH -15 NA  3\n",
       "4     0   25M TAGTGGACGAGTCCGGAATCGAACC HHHEGGGGGGGGGFFFFFCCBBACB -10 NA  2\n",
       "5     0   25M TAGTGGACGAGTCCGGAATCGAACC HHHEGGGGGGGGGFFFFFCCBBACB -10 NA  2\n",
       "6     0   25M GGTTCGATTCCGGACTCGTCCACTA BCABBCCFFFFFGGGGGGGGGEHHH -10 NA  2\n",
       "7     0   25M GGTTCGATTCCGGACTCGTCCACTA BCABBCCFFFFFGGGGGGGGGEHHH -10 NA  2\n",
       "8     0   25M TAGTGGACGAGTCCGGAATCGAACC HHHEGGGGGGGGGFFFFFCCBBACB -15 NA  3\n",
       "9     0   25M TAGTGGACGAGTCCGGAATCGAACC HHHEGGGGGGGGGFFFFFCCBBACB -10 NA  2\n",
       "10    0   25M TAGTGGACGAGTCCGGAATCGAACC HHHEGGGGGGGGGFFFFFCCBBACB -15 NA  3\n",
       "11    0   25M GGTTCGATTCCGGACTCGTCCACTA BCABBCCFFFFFGGGGGGGGGEHHH -10 NA  2"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapped_reads = untreated\n",
    "anticodon = \"AGC\"\n",
    "tRNA_annotations = tRNA_annotations\n",
    "anticodon_tRNAs = filter(tRNA_annotations, Anticodon %in% anticodon)\n",
    "tRNA_read = merge(anticodon_tRNAs, mapped_reads, by = c(\"Chr\",\"Start\",\"End\"))\n",
    "x = tRNA_read\n",
    "\n",
    "#head(filterMature(x %>% filter(qname = \"SRR1874045.362855\"), clean = T, flip = F))\n",
    "\n",
    "x %>% filter(qname == \"SRR1874045.292256\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp = assignReads(untreated, tRNA_annotations, anticodon = \"GCC\", useBam = F, returnCounts = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp2 = assignReads(untreated, tRNA_annotations, anticodon = \"GCC\", useBam = T, returnCounts = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp3 = merge(data.frame(table(tmp$read)), data.frame(table(tmp2$read)), by = \"Var1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>score</th><th scope=col>read</th><th scope=col>chr</th><th scope=col>tRNA.start</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrIV</td><td>410379</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrVI</td><td>204924</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrVII</td><td>774349</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrVIII</td><td>146242</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrX</td><td>197313</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrXI</td><td>219895</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrXI</td><td>517988</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrXII</td><td>656934</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrXIII</td><td>321147</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrXIII</td><td>768369</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>1.990687</td><td>SRR1874045.292256</td><td>chrXVI</td><td>856902</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llll}\n",
       "  & score & read & chr & tRNA.start\\\\\n",
       "\\hline\n",
       "\t1 & 1.990687 & SRR1874045.292256 & chrIV & 410379\\\\\n",
       "\t2 & 1.990687 & SRR1874045.292256 & chrVI & 204924\\\\\n",
       "\t3 & 1.990687 & SRR1874045.292256 & chrVII & 774349\\\\\n",
       "\t4 & 1.990687 & SRR1874045.292256 & chrVIII & 146242\\\\\n",
       "\t5 & 1.990687 & SRR1874045.292256 & chrX & 197313\\\\\n",
       "\t6 & 1.990687 & SRR1874045.292256 & chrXI & 219895\\\\\n",
       "\t7 & 1.990687 & SRR1874045.292256 & chrXI & 517988\\\\\n",
       "\t8 & 1.990687 & SRR1874045.292256 & chrXII & 656934\\\\\n",
       "\t9 & 1.990687 & SRR1874045.292256 & chrXIII & 321147\\\\\n",
       "\t10 & 1.990687 & SRR1874045.292256 & chrXIII & 768369\\\\\n",
       "\t11 & 1.990687 & SRR1874045.292256 & chrXVI & 856902\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "Source: local data frame [11 x 4]\n",
       "Groups: read [1]\n",
       "\n",
       "      score              read     chr tRNA.start\n",
       "      (dbl)             (chr)   (chr)      (chr)\n",
       "1  1.990687 SRR1874045.292256   chrIV     410379\n",
       "2  1.990687 SRR1874045.292256   chrVI     204924\n",
       "3  1.990687 SRR1874045.292256  chrVII     774349\n",
       "4  1.990687 SRR1874045.292256 chrVIII     146242\n",
       "5  1.990687 SRR1874045.292256    chrX     197313\n",
       "6  1.990687 SRR1874045.292256   chrXI     219895\n",
       "7  1.990687 SRR1874045.292256   chrXI     517988\n",
       "8  1.990687 SRR1874045.292256  chrXII     656934\n",
       "9  1.990687 SRR1874045.292256 chrXIII     321147\n",
       "10 1.990687 SRR1874045.292256 chrXIII     768369\n",
       "11 1.990687 SRR1874045.292256  chrXVI     856902"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp %>% filter(read == \"SRR1874045.292256\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>score</th><th scope=col>read</th><th scope=col>chr</th><th scope=col>tRNA.start</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>-10</td><td>SRR1874045.292256</td><td>chrVI</td><td>204924</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>-10</td><td>SRR1874045.292256</td><td>chrVIII</td><td>146242</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>-10</td><td>SRR1874045.292256</td><td>chrX</td><td>197313</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>-10</td><td>SRR1874045.292256</td><td>chrXI</td><td>219895</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>-10</td><td>SRR1874045.292256</td><td>chrXI</td><td>517988</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>-10</td><td>SRR1874045.292256</td><td>chrXIII</td><td>321147</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>-10</td><td>SRR1874045.292256</td><td>chrXVI</td><td>856902</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llll}\n",
       "  & score & read & chr & tRNA.start\\\\\n",
       "\\hline\n",
       "\t1 & -10 & SRR1874045.292256 & chrVI & 204924\\\\\n",
       "\t2 & -10 & SRR1874045.292256 & chrVIII & 146242\\\\\n",
       "\t3 & -10 & SRR1874045.292256 & chrX & 197313\\\\\n",
       "\t4 & -10 & SRR1874045.292256 & chrXI & 219895\\\\\n",
       "\t5 & -10 & SRR1874045.292256 & chrXI & 517988\\\\\n",
       "\t6 & -10 & SRR1874045.292256 & chrXIII & 321147\\\\\n",
       "\t7 & -10 & SRR1874045.292256 & chrXVI & 856902\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "  score              read     chr tRNA.start\n",
       "1   -10 SRR1874045.292256   chrVI     204924\n",
       "2   -10 SRR1874045.292256 chrVIII     146242\n",
       "3   -10 SRR1874045.292256    chrX     197313\n",
       "4   -10 SRR1874045.292256   chrXI     219895\n",
       "5   -10 SRR1874045.292256   chrXI     517988\n",
       "6   -10 SRR1874045.292256 chrXIII     321147\n",
       "7   -10 SRR1874045.292256  chrXVI     856902"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp2 %>% filter(read == \"SRR1874045.292256\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.998396979572411"
      ],
      "text/latex": [
       "0.998396979572411"
      ],
      "text/markdown": [
       "0.998396979572411"
      ],
      "text/plain": [
       "[1] 0.998397"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cor(tmp3$Freq.x, tmp3$Freq.y, method = \"spearman\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp4 = merge(tmp, tmp2, by = \"read\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#head(tmp4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#tmp5 = filter(tmp4, read %in% names(which(table(tmp4$read) > 1))) \n",
    "#tmp5[with(tmp5, order(-score.x, read)), ][1:19,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtAAAAJYCAMAAAB8aiEbAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3deWATdf7/8XdobWmhUMpRrip3qVBEERAFEW/QyvVlYdGt8pPuoWz97q712N3qb1cBd1fXVcQDbxd0F+8LRUS8ADkKokC55SqlFFBu6JH5ZpK0TTJJ5j2T9+RoXs8/2iQz6afZPLZOmzcTUhBqRFGkvwGEJANo1KgCaNSoAmjUqAJo1KgCaNSoAmjUqAJo1KgyD/rowSBV1wTbKtKxE5YvUV1r+RJHrX8UVXbLlzhy0vIlztgPBdy2ok/b70MHfaQySDW1wbaKdOyE5UtU2y1f4uhJy5eoVixf4sgpy5eoUg4G2vRyMyqsAWhGAM0skqAPFNmSHq2sZwnQQQJoZhEEvXsMtV1QCdCsAJpZ5ECvzaXz1lUCNC+AZhYx0Avb0/W71AsAzQmgmUUK9BNJTYoPOC8BNCeAZhYZ0BUFlPK8+zJAcwJoZhEBvf0Kyvq87gpAcwJoZpEAXZJDQzbXXwNoTgDNLAKg38qgSWUNVwGaE0AzCz/oGYkJ0z2vAzQngGYWbtD78qnlfK/NAM0JoJmFGfSmi6jnCu/NAM0JoJmFF/RXXWj4Fp/NAM0JoJmFFfQrzalwv+9mgOYE0MzCCdo5XKcJoDkBNLPwgd4zltp+6GczQHMCaGZhA73+Quq3zt9mgOYE0MzCBfqT9nTdTr+bjYGumqx+nJnn6EGAFg2gmamgZyfb3MN1moyAPvPtQ3nq59vnlJSUbANo0QCaWZVyoNCW8lygzUZAf3zrzSro2rEbvW4GaIkAmlnVkSup85KAm42AVpQ9KujyvBmTbn7mlHr90CJHO48GqdYebKtIp05bvkQYHsXJM5YvUaNYvsTJKsuX2JpLF24JuPWICdCr8l7curzgMfX6qgGOFnPuiZBIn7WmW04H3lx/FgMDoM8cdXxYP7rK8bH8JUebjgfJbg+2VaQzVZYvUatYvsTpxvEoqi1e4JHEhH+dCLL9mAnQzk7l1R+t4BhaIhxDM9qXTy3eD3yiGTUToN99wK4omyfYAVoygNZv0xDq8U2QMyepmQC9c/STpSt/Pa/+RoCWCKB1+7oLXbol2KnA1EyAVjbeM7FgfsPRN0BLBNB6/bs5FZYHPbedmjHQfgNoiQBap+KEpEfUzwAtEEAzswz0nnHU5gPnJYAWCKCZWQV6w0DKWe26CNACATQzi0B/0oFG1Q3XAbRAAM3MGtBPJtuK6ofrAFoggGZmBeiKQlvKsw1XAVoggGZmAegdV1HnzzyuA7RAAM1MHvSac2nA9543ALRAAM1MHPQ7GTSxzOsWgBYIoJlJg/5XUpPpPjcBtEAAzUwWtDpc9x/fGwFaIIBmJgp66wjq8Y3mVoAWCKCZSYJe2pWG+Z65rhKgRQJoZoKg5zangnI/twO0QADNTA50ccJZj/jdANACATQzKdB7x9cN12kCaIEAmpkQ6A0DqffqANsAWiCAZiYDepHHcJ0mgBYIoJmJgH4x1WO4ThNACwTQzARAVxTams4Jsh2gBQJoZqGD3pVHnT4LtgNACwTQzEIGveZcuuD7oHsAtEAAzSxU0O9m0MS9wXcBaIEAmlmIoB9PahLoROb1AbRAAM0sJNDl+dTiNd29AFoggGYWCuhtI+jsL/V3A2iBAJpZCKCXdaNhmxn7AbRAAM3MPOjX02mqv+E6TQAtEEAzMw064HCdJoAWCKCZmQRddhO1fp+5L0ALBNDMzIHeOIiyV3F3BmiBAJqZKdCfdqSRP7D3BmiBAJqZGdAvplKQ4TpNAC0QQDMzDlpvuE4TQAsE0MwMg959A2V+ZOgeAC0QQDMzCnpNHzo/+HCdJoAWCKCZGQT9XmuaoDNcpwmgBQJoZsZAz2IM12kCaIEAmpkR0PsLKO1V40sAtEAAzcwA6G2Xs4brNAG0QADNjA96WTcayhmu0wTQAgE0MzboN1oxh+s0AbRAAM2MC5o/XKcJoAUCaGY80GW/oPTXzS4B0AIBNDMW6NJBlL3S9BIALRBAM+OAXtyJruUP12kCaIEAmhkD9EvNqHB/CEsAtEAAzUwX9IEiW9NnQloCoAUCaGZ6oHePpnYLQlsCoAUCaGY6oNf2pf7fhbgEQAsE0MyCg/440/hwnSaAFgigmQUFbWq4ThNACwTQzIKArjA3XKcJoAUCaGaBQW+/grK+kFgCoAUCaGYBQS/vTpeYGq7TBNACATSzQKDfbEWTy2SWAGiBAJpZANDTE00P12kCaIEAmplf0PvyKX2+2BIALRBAM/MHunQw9TI/XKcJoAUCaGZ+QH95Dl0TwnCdJoAWCKCZaUG/HOJwnSaAFgigmfmCPlBkS35adgmAFgigmfmA3jMm5OE6TQAtEEAz8wb9bW7ow3WaAFoggGbmBXphe7p+l/gSAC0QQDPzBP2EyHCdJoAWCKCZNYCuKLSlvmDFEgAtEEAzqwe9/Uqh4TpNAC0QQDOrA12SQxfLDNdpAmiBAJqZG/RbGWLDdZoAWiCAZuYCPTMxUWy4ThNACwTQzFTQ+/KppdxwnSaAFgigmTlAb7qIeq2wcAmAFgigmR059VUXunqHlUsAtEAAzezI/ObCw3WarAd9/Kcg1dqDbRXp5CnLl6ix/lGcOG35EjWK5Uv80Zb0uMVLVCtHgm0WAH0qWHZ70M0SVVVbvoQ9+IOUKAyPotbqR3F4AmUusXgNnUdxQgA0DjkkagSHHOsvpP7brF2iEsfQIgE0o0/a0/Xlht+83nAALRBA6/dkcpPiA4bfvN54AC0QQOtVUWhLed74m9ebCKAFAmiddlxFWZ9XArRIAM3MOtAlOTTEOVwH0AIBNDPLQL+dQZNcw3UALRBAM7MK9MzEhOnuiwAtEEAzswa013AdQAsE0MwsAb31MurZMFwH0AIBNDMrQH/dlYZvabgK0AIBNDMLQM/1Ga4DaIEAmpk86OKEpEe9bgBogQCamTToPeOo7YfeNwG0QADNTBj0hoHUb53PbQAtEEAzkwW9qANdt9P3RoAWCKCZiYJ+Ktnm58x1AC0QQDMTBK0O1z3n53aAFgigmcmB3nE1dV7ibwNACwTQzMRArzmXBqz3uwWgBQJoZlKg36kfrtME0AIBNDMh0I8l1Q/XaQJogQCamQjo8nxq8d+AWwFaIIBmJgF66wjq8U3gzQAtEEAzEwC9tCtduiXIdoAWCKCZhQ56XhoVlgfbAaAFAmhmIYMuTkjSOZE5QAsE0MxCBF12I7X5QGcfgBYIoJmFBnrjQMpZrbcTQAsE0MxCAv1pRxqlGa7TBNACATSzUEC/mGorYrwtLEALBNDMzINWh+ue5ewI0AIBNDPToHflUefPWHsCtEAAzcws6DV9aMD3vF0BWiCAZmYS9LsZNJH7trAALRBAMzMH+vGkJgGH6zQBtEAAzcwMaHW47j/83QFaIIBmZgL0tuDDdZoAWiCAZmYc9LJuNCzYcJ0mgBYIoJkZBv1GOhUEHa7TBNACATQzo6CLE87SGa7TBNACATQzY6DLbqI27xtdAqAFAmhmhkBvHES9dYfrNAG0QADNzAjoTzvSSP3hOk0ALRBAMzMA+qVmrOE6TQAtEEAzY4M+UGhrOsfUEgAtEEAz44LefQN14g3XaQJogQCaGRP02r50AXO4ThNACwTQzHigP8qkiXvNLgHQAgE0MxboWUlN/JzInBtACwTQzBig9xdQi9dCWAKgBQJoZvqgt11OZ38ZyhIALRBAM9MFvbw7Ddsc0hIALRBAM9MD/WYrmmpsuE4TQAsE0Mx0QBcnGh6u0wTQAgE0s6Cgy35BrQ0P12kCaIEAmlkw0KWDKXtV6EsAtEAAzSwI6MWdaOQPAksAtEAAzSww6JebkanhOk0ALRBAMwsE+kCR2eE6TQAtEEAzCwB69xjK/EhoCYAWCKCZ+Qe9NpfONztcpwmgBQJoZn5Bf5xJE0wP12kCaIEAmpk/0E+ENFynCaAFAmhmWtAVBZT2quQSAC0QQDPTgN5+RYjDdZoAWiCAZuYLenl3GhracJ0mgBYIoJn5gH4r9OE6TQAtEEAz8wY9Q2C4ThNACwTQzDxB78un9NfllwBogQCamQfo0osoe6UFSwC0QADNrAH0V+fQNRLDdZoAWiCAZlYP+pVmVLjfkiUAWiCAZuYGrQ7XPWPREgAtEEAzc4HeM5baLbBqCYAWCKCZOUF/m0v9v7NsiVgDXTVZ/Vg7b8otc2sBWrQwgV7YXnK4TlNsgT7z7UN56ucF+atX5y8AaNHCA3p2suhwnabYAv3xrTc7Qd+2SFEW3Q7QooUDdG2hTXa4TlP0gj5858VXPF+jXrovu+HWPSromhsqFKXiBnXj6b2OKg8HqbY22FaRTpy0fIkau+VLHD9l+RIHR9HZX1m7xLHT1n59R1XKj0G2HgoA+kAXcnSTeqD8K/IBXZl3xkE5T73rqgGOFuv/bEdR0I5cuqxSf7fYrqb+kjfo2+mhyhXX0mwlMGj1f5vtdztaezpIdnuwrSJV11i+hF2xfAnLH8XCDPp/Ry1e43SV9c9FbdDn4lQA0D0mOD6cGNDykB/QVarlirzquhtxDC2R1cfQMxPPesr4m9cbLWqPoVP+pn4ssRX5Aa1MW6Ion02rvxGgJbIWtDpcN9/4m9cbLmpB9/+581NB4go/oD+4ddPGKR8CtGiWgt40hHqtNP7m9caLWtC32WYfd3z6qXPXUi1o+9wpU161A7RoVoL+qotzuC6eQe/rSk3+5fi8qo0tQ+8v1AAtkYWg/93cNVwXz6CVw38f/Zz6ecdtHQA6tkEXJyQ/7bwQ16A9OgnQMQx6z7j64bp4B318+qD0s5Q7SnU8A7RIFoHecGHDcF2cgz6US1mXktIpQ080QEtkDehPOtD1u+quxDnoQnqyppSUb9vcCNCxCvpJr+G6OAfd+SZFcYBWpmYBdGyCrii0pb7gcT3OQTef6QI9MwWgYxL0jqso6wvPG+Ic9NBhNU7QebkAHYug1+TQxd5nrotz0M9Q3jcr6aeZ9DuAjkHQ72TQ5DLvm+IctH0qOet0GKBjD/RDiYmaM9fFOWhF+XxM75SeBQd1PAO0SKKg9+VTy/maW+MeNDOAlkgS9NbLqNcK7c0ADdDuYgv00q509Q4/t8czaLpbofoAOqZAz20e4Mx18Qw6+2Eluz6AjiXQxQlJj/rfEs+gjQTQEgmB3jOe2gY6cx1AA7S7mAG9YSCdty7QxjgHTUucn+yPpAF0rIBe5DlcpymeQd/Y8Csh9QPoGAH9VPAz18Uz6DuzsynL9Sth7jyAjgnQFYW2lOeD7RDPoJX6Qw79AFqikEHvyqOsz4PuEeeg+9z56WmAdhUDoNecS0N03hY2zkGfQ5R6/RNbAboyFkC/m0GTynT2iXPQys6Xb+1B1P329wE66kE/lpQwXXeneAetVjanB176jnrQ5X6H6zTFPehD7/1hYAI1uwagoxv0thHU089wnaY4Bz2tn43SRv3tmyodzwAtknnQS7vS8C2cHeMcNNG586p9bwRoqzINel4a921h4xz0PZckUduxj6zAT+ioBh14uE5TnINWlFNfTr+2BaWMAOioBV12I7X9kLtz3INWlNrvHu2Jv3JEL+iNA6lfwOE6TXEOuvb7WeNaE/W/ewlARynoTzvSdTv5u8c56DZEbW98Zb+OZoAWygToF1Ntht4WNs5BXzp9da3vbQBtWYZBq8N1zxm6R5yDVuxv3zJ0cP7rds0GgLYgo6B330Cdlxi7S5yDrrraNd9/ld7f7QBaIoOg1/ShAesNLhHnoP9C45Ye/mnZWHoQoKMO9HuM4TpNcQ66/5XOQ+jayy8A6GgDPYszXKcpzkGnPuT6PLMZQEcX6P0F1OK/JpaIc9A5E12ff9YHoKMK9LYR1OMbM0vEOejf0r2HFOXQvVQI0NEEelk3upQ1XKcpzkEfzSZq354o+yhARxHoN1pRYbm5JeIctHLiwYGtWg188ISOZ4AWiQm6OCFJcyJzbvEOmhtAS8QCXXYTtfnA9BIArawYNfiOSn8bAFo6DuiNgyhntfkl4hr0u+enXFSyuyURdTwA0NEBenFHGmVguE5TPIP+xPELYWKbXw5Yd/hh+hVARwXol5rZiowM12mKZ9DDU5crhy+mNY6LV/cC6CgAfaDQlvJsaEvEM+g2+Y4PH9IZx8e7EgE68qB3j6bOn4W4RDyDpj85PnznvO1+/BOsyINe25cGfB/qEnEN+n7F9U7fAK0WadAfZ9JEw8N1mgAaoN1FGPSspCYmhus0ATRAu4soaHW47j8SS8Q1aM8AOpKgt19ucrhOUzyDzvYMoCMIenl3GmZuuE5TPIM2EkBLFAD0m62owORwnSaABmh3EQNdnHiW6eE6TQAN0O4iBHpfPrV5X24JgAZod5EBXTqYeocwXKcJoAHaXURAf3kOjQxluE4TQAO0u0iAfjnU4TpNAA3Q7sIP+kCRrekc4SUA2tkl2bljVwK01XmD3j2GOoU6XKcJoJ1l0sm3zwZoq/MCvTaXLgh5uE4TQPMCaIk8QS9sTxP3yi8B0ADtLrygn0hqYuhE5tziHLT9sfGKsnj45a8CdFhBVxRQi9csWSLOQc+mTkp5cyJ6E6DDCHr7FXT2l9YsEeeg+/UqUx6nZ3Z3GwrQ4QNdkkPDNlu0RJyDbnaXoozuVKvckwHQYQP9VgZNlRqu0xTnoFveo1S1vklR/tIcoMMFeobkcJ2mOAc98LwzL9O/ldpBfQE6PKD35VNrweE6TXEOeg6lU8vDO7rS3wE6LKA3XUTZq6xcIs5B1/6pXae3lVVn330GoMMBuqQLjfzB0iXiHDS7oweDVFsbbKtIx09YvkSN3fIl/tOc7qq0dolqxdqv7+joKcuXqFIOBdssAPpMdZDsSrCtItXWWr6E9Y+i2JYy1+o1wvBc1ET6uWg4nNCArjuJQUqXgQ8eCgYahxwht2cstf/c2iUq4/6QI7stUbv2RJkdkykz2DmiATrU1l9I5282/ub1Rotz0Fuy/uw4IDlc3HNv9YKMAoC2rk/a04S9ht+83nhxDvqan7s+TxqtKM8F+1s0QIfW7GR1uA6gmZkG3aLunWRbKcrKFIC2qIpCW9qrlYbfvN5McQ76/Cvc7/XdR1FmdQNoa9pxpXu4DqCZmQb9Nxq77Keflo+jB4/9o/k0gLakkhwa6hquA2hm5l8p/Lnrz3bXVZXSefsB2orebhiuA2hmIbxSuPjWYYPz37IrB985GcQzQJtupsdwHUAzw78pFMgS0PvyKf31+msAzcw06EPjU3HCc3dWgN40hLJXNlwFaGamQd9KTXvihOeuhEHf27yy8usudI3ncF0Ugf5hv+klohl0+2FHdCQDtKkS1f/q2ZpToZebqAH993aUlGf2LDfRDLrtMzzPAG2sHPeB3DPeN0cL6H84v7lck28iF82gb/klQNclCdrmBn2v981RAnp/huu7m2VuiWgGfXzCnZv0/rEKQBuvbiw30fvmKAG93v3d/dbcEtEMGm/r1pAVP6Fbed8cJaB3NnF9d38yt0Q0g8bbujUkCdothnxujhLQldc5v7mmK8wtEc2g2QG0gSoK3T+hz/LZEC2gN3Z3fHNJ/zS5BEALFEugd1xFWV8k2Mg2zHdLtICuLJv1u/+/zOwSUQua7sYxtEdSoNecS5cEOHNd1IAOqagFnf0wjqE9EgL9TgZNDvQHXoBmhkMOgWRA/yspyJnrAJqZedD296YMG5z/hh2gRUCrw3XzA28GaGamQVdd4zqCvqoKoAVAbx1BvVYG2Q7QzEyD/jNd+/mhH7+4hooBOnTQS7t6D9dpAmhmpkH3vbha/VR9US5Ahwx6ru9wnSaAZmYadPoDrs9/bQHQoYIuTkh+WmcXgGZmGvTY8a7P44YDdGig946ndgv0dgJoZqZBVwwqcmytvLPZaoAOCfSGgdT/O929AJpZSNN2mZlEaeprK38FaLMt6kDX79LfDaCZmQbds9s5nTu0a52elpqcaLsdoE32YirvbWEBmhleKRTINOiKQlvqC6w9AZoZXikUyCzoXXmU9QVvV4BmhlcKBTIJes25dDH3bWEBmhleKRTIHOh3gwzXaQJoZnilUCBToB9LSjTwtrAAzQyvFApkAnR5PrUMMlynCaCZ4ZVCgYyD3jaCehn6h6YAzQyvFApkGPSybnT1DkP3AGhmeKVQIKOgX0/XG67TBNDM8EqhQAZBFyckPWp0CYBmFvIrhUe3B8EM0NrKbqK2usN1mgCaWcig/5kO0EZAbxxE560zvgRAMzMN+uCoZNdLhTiNgRHQn3ZkDddpAmhmpkH/hto2p17dE7IWAjQfNHe4ThNAMzMN+uwh1fts25WSNhsAmgu6otCW8ry5JQCamWnQSX9RlAvmKsrvRwI0E/TuGyjrc5NLADQz06A7FCjK1BsVZTZ+KWSCXtOHhnCH6zQBNDPToIe2ml/zRLPy2pGdAZoF+r3WNMnkW5NUAjQ706DnER3b3DSlLd0G0BzQs5ISpoewBEAzM/9e33dkHFNe799y8k8ArQ96f4Gx4TpNAM0M/6ZQIF3Q2y6nnibfxcEdQDMDaIH0QC/rRsO3hLYEQDMzDXrHlSk4g787HdBvtDI8XKcJoJmZBn0t2brjDP6ugoM2M1ynCaCZmQadNv6QjmSAViv7BbX9MPQlAJqZ+XnoWTzPcQ66dBD1MzFcpwmgmZkGPafLGoB2Fxj04k503U6JJQCamWnQ1QOpPY6hXQUE/VIzm7nhOk0Azcw06P/F+xTWFwD0gSJbynNCSwA0M9Ogu/fbUKtDOc5B7x5NnZdILQHQzEyDbv00i3P8gl7blwasF1sCoJmZBv2zPwB0Xf5Af5wZynCdJoBmZv7fFA6ezftDdHyCDnG4ThNAMwvpRDP4pdCVBnRFAbX4r+gSAM3MNGi8eX1DvqC3X0E9vpFdAqCZYdpOIB/Qy7vTpSEO12kCaGYALZA36DdbUWG59BIAzQygBfICPT0xycCJzLkBNDOAFsgD9L58avOBBUsANDOAFqgBdOlgylltxRIAzQygBaoH/eU5NEpkuE4TQDMzDbpU/XPdU4cBurIB9MvNbEUyw3WaAJqZOdBd8+6d2dPuuLUUoCvrQKvDdc9atQRAMzMHumhkFlGr4dPorVMA7Qa9ewx1/syyJQCamelDjjX06M3nESX0zQdoFfS3uTTge+uWAGhm5kCvO62UqrfQa89OG+EreGaeowfjB/SGwS1b5L7QniYKDtdpAmhm5kBTQs6V9EG5/2Po2+eUlJRsixvQPzR1jmg1ER2u0wTQzMyBXvH874Y5nsSOdPv8zTU+nmvHboyrQ44RrpnDTtauAtDMzP/ZrufWN+6j1kRpPhvK82ZMuvkZ5y+La0c4+sIeJEUJtjUmynCBPivS30fINYLnwq73KKqDgFb/Dk2lB5f4np9jVd6LW5cXPKZe/O4GR0trgqQowbaKVGu39uu3coO2dhWrH4Uje+w/FzV6j6IqMGhn4/dqbztz1PFh/ej6uzb2Q46+LtDdrF0FhxzMLHrp+1Re/V0bOeiZCa7fCb+2dBWA5iYP+t0HHMcxmyfY4wK0Olz3ivpnO+F/oKIJoJnJg945+snSlb+eV3+9MYPedBH1Xm38zetNBNDMLDjk2HjPxIL5DX/Ma8Sgv+pCI9XhOoBmFpugfWq8oF9p7h6uA2hmAC2QZaCLbE3nuC4BNDOAFsgi0HvGUqe64TqAZgbQAlkDev2FdEH9cB1AMwNogSwB/Ul7mri3/hpAMwNogawA/WRyE88TmQM0M4AWSB50RaGtxWueNwA0M4AWSBz0jqvo7C+9bgFoZgAtkDTokhwautn7JoBmBtACCYN+O4Om+p65DqCZAbRAsqBnJp6lPXMdQDMDaIEkQe/Lp9bva28GaGYALZAg6K2XUfYqP7cDNDOAFkgO9NddaeQP/jYANDOAFkgM9NzmFODMdQDNDKAFkgJdnFA3XKcJoJkBtEAyoPeMo8yPAm0EaGYALZAI6A0D6fzAZ64DaGYALZAE6EUdaMLewJsBmhlACyQA+inv4TpNAM0MoAUKGXRFoS3t1aB7ADQzgBYoVNA7rvYdrtME0MwAWqAQQG98dnPlmnM1w3WaAJoZQAtkGvQnzYgoLV07XKcJoJkBtEBmQe8/y3UWRsbbwgI0M4AWyCzo37o80/P6uwI0M4AWyCzoS9ygf66/K0AzA2iBzIKe6AZ9n/6uAM0MoAUyC/oJl+cm+/V3BWhmAC2QSdDFCYk2h2fbM4x9AZoZQAtkCnTZjdRuwfejeo/1O9DvG0AzA2iBzIDeOJD6f8ffHaCZAbRAJkB/2jHocJ0mgGYG0AIZB/1iavDhOk0AzQygBTIKWh2um2fsLgDNDKAFMgh6Vx5lfWFwCYBmBtACGQO9pg9dojdcpwmgmQG0QIZAv5tBk8sMLwHQzABaICOgH0/yc+Y6/QCaGUALxAddnk/p880sAdDMAFogNuhtI6jXSlNLADQzgBaIC3pZN7qG9UK3NoBmBtACMUG/kU6FjME6vwE0M4AWiAe6OCH5adNLADQzgBaIA7rsJmq3wPwSAM0MoAVigN44yNBwnSaAZgbQAumD/rQjXb8rlCUAmhlAC6QL+qVmBofrNAE0M4AWSAf0gUJb6gshLgHQzABaoOCgd99gfLhOE0AzA2iBgoJe25cuNjxcpwmgmQG0QMFAf5RpZrhOE0AzA2iBgoCelZRoZrhOE0AzA2iBAoLeX0AtTQ3XaQJoZgAtUCDQ2y6nXitklgBoZgAtkB/Ql6dn3ru8O129I/g9tz/2+9l7OEsANDOAFkgD+nv1FF9k0x2u+7CtY7esrxlLADQzgBZIAzrVdRbG83Tut7Ojc7feOFmjXAAtkAa0+zS5STr3m+feb6H+EgDNDKAFCgQ6Qed+s937Mf4QAtDMAFqgQKBb6txvkfv80Ov0lwBoZgAtkC/olxNcUt/Qu2Oec7epjCUAmhlAC+QN+kCRLaWL+hn4f0oAAA5mSURBVEeOu3Xv+MPUZGr2B84r4wDNDKAF8gK9ewxlfe74Kb2Eddfyb3n/bBagmQG0QJ6g1/alIaEP12kCaGYALZAH6I8zaZLAcJ0mgGYG0AI1gH4iKWG6JUsANDOAFqgOdIXYcJ0mgGYG0AK5QW+/gnoKDddpAmhmAC2QC/Ty7jR8i1VLADQzgBbICfqtVubPXKcfQDMDaIFU0DMSkx61cAmAZgbQAh07sS+f2n5o5RIAzQygBTq28yLqx5gwCiGAZgbQAq3sStfttHYJgGbWKECfqQmSogTbKtFbzW0zqi1ew275o6iptVu+RDw8iqpY/wl9oMiW+oq1S1TiJzS7RvETOpKg94ylzkuNv3m90QCaGUCH1re5NGC98TevNxxAMwPokFrYXh2uA2hmAB3loGcnO4frAJoZQEc16IpCW4v/qhcAmhlARzPo7VdSj2+clwCaGUBHMeiSHLrUPVwH0MwAOnpBv51BheXuywDNDKCjFvTMxKSGE5kDNDOAjlLQ+/KpzQcNVwGaGUBHJ+hNQyhntcd1gGYG0NEG+m7nORi70Civ4TqAZgbQUQZ6qfssjEXebwsL0MwAOspA29ygfW4GaGYAHWWg3Z6pwPtmgGYG0FEK+h/eNwM0M4COMtA45AgtgI4u0E8muzyn+dwO0MwAOppAq8N1c9Sf0ef4bgFoZgAdRaB3XFU3XKcJoJkBdPSAXpNDwwKduQ6gmQF01IB+J4MKygNtBGhmAB0toB9KPOuRwFsBmhlARwdodbju/SDbAZoZQEcF6K2XUe/VwXYAaGYAHQ2gv+5KI4OfuQ6gmQF0FICe29zmM1ynCaCZAXTkQRcnNJ2jtw9AMwPoSIPeM546faa7F0AzA+gIg94wkC74Xn83gGYG0JEFvagDTdzL2A+gmQF0REE/ldykWOfXQVcAzQygIwhaHa57jbcrQDMD6MiB3pVHZ3/J3BegmQF0xECvOZeGbubuDNDMADpSoN/NoKkBh+s0ATQzgI4Q6MeSgg3XaQJoZgAdEdDl+dQ62HCdJoBmBtCRAL1tBGWvMnQPgGYG0BEAvbQrjfzB2F0AmhlAhx/0vDTfM9fpB9DMADrsoIsTmj5jZH9nAM0MoMMMuuxGyvyIv3tdAM0MoMMHOqlJQuXGgXQ+Y7hOE0AzA+hwgc50n7RuAme4ThNAMwPocIF2e25r9NdBVwDNDKDDBDrJDdpm7JHVBdDMADpMoJsANECziwHQiQAN0OxiAPR7btApxh5ZXQDNDKDDA/px9zF0E4OPrC6AZgbQ4QC9v4DSX69sYrP1NvzQ3AE0M4AOA2h1uG6l8UflGUAzA2jrQS/rRtcYHK7TBNDMANpy0G+0osL9Zh6VZwDNDKCtBm1quE4TQDMDaGtBl91E7RaYfFSeATQzgLYU9MZB1P87s4/KM4BmBtBWgl7c0eRwnSaAZgbQFoJ+qRnzzHX6ATQzgLYM9IFCW9q8UB6VZwDNDKCtAr17NGV9EdKj8gygmQG0RaDX9qVL2Geu0w+gmQG0NaA/zqTJZSE+Ks8AmhlAWwJ6lrEz1+kH0MwA2gLQ6nDd/NAflWcAzQyg5UFvv5x6hThcpwmgmQG0OOjl3UMfrtME0MwAWhr0mxLDdZoAmhlAC4MuTkx+SuZBeQXQzABaFPS+fJnhOk0AzQygJUGXDhYartME0MwAWhD0l+fQ9bvkHpVnAM0MoOVAvyw3XKcJoJkBtBToA0W21BdEH5VnAM0MoIVA7x4jOVynCaCZAXSAaudNuWVuLRv02ly6WHC4ThNAMwPoAC3IX706fwEX9ML2ssN1mgCaGUAH6LZFirLodibol5ISZYfrNAE0M4D2X80NFYpScUON4+Lm3zhaXRW404WU/kGQ7RLV1Fi8QFWVXbF8iZpay5cIw6OojvSjOG0CdGXeGUU5nXfIcXHVAEeLA+/600jK2cL+wgiFWk39JcOg1R/uNUccHT4YqJIcuu7HgFulOn7S8iVq7JYvceyU5UtUK5YvcfS05UtUKYeCbTYBukq1XJFXXXc94DH0WxlUeMbwm9cbDsfQzHAMHaBpSxTls2n1VwOBnpGY9KjxN683HkAzA+gAfXDrpo1TPtQBvS+f2i4w/ub1JgJoZgAdIPvcKVNetQcHvekiOm9dJUCzA2hmkXnp+6su7uE6gGYG0MwiAvqV+uE6gGYG0MwiAbrIlvK8+yJAMwNoZuEHvWcsZX1edwWgmQE0s7CDXn8hDWkYrgNoZgDNLNygP2lPkzyG6wCaGUAzCzPo2ckJ0z2vAzQzgGYWVtAVhbaW3meuA2hmAM0snKB3XEk9V3h/fYBmBtDMwgi6JIeGb/H5+gDNDKCZhQ/02xl+zlwH0MwAmlnYQM9Uh+s0ATQzgGYWJtDqcN2Hfr4+QDMDaGbhAb1pCPVb5+/rAzQzgGYWFtBfd6Hrdvr9+gDNDKCZhQP0v5vbAp25DqCZATSzMIAuTkh5LtDXB2hmAM3MetD3Ov+tVYAAmhlAM7Me9PjctYG//meLpB+Qpp+OWr7E5wstXyIMj+IrS95Awasfj1m+xNIFFcE2C4AuPx5kY95Vpr9uFDVxaKS/A4mmDLDr7xT13TbgBGs/86CDBtDRE0ALBNDRE0ALBNDRE0ALtK1RnKtx+6ZIfwcS/bAx0t+BRLs21urvpFgGGqHIBNCoUQXQqFEF0KhRZQlon/fLitFm5jl6MNLfRWhVTVY/xvrz4XoUvOfDEtA+75cVo90+p6SkZFukv4tQOvPtQ3nq59h+PuoeBe/5sAS0z/tlxWa1Y2P+z10f33qzk0JsPx/uR8F8PqwA7fF+WTFced6MSTc/cyrS30Zo7VEpxPzz4XwUzOfDCtAe75cVw63Ke3Hr8oLHIv1thJaTQsw/H85HwXw+LARdqb9nNHfmqOPD+tFVkf4+QsoTdOw+H85HwXw+rADt+35ZMdypGGag5qQQ88+H81E4038+LPml0Of9smKzdx+wK8rmCbE92OOiEOvPh/NRMJ8PS0D7vF9WbLZz9JOlK389L9LfRmi5QMf68+F8FMznwxLQPu+XFaNtvGdiwfzY/dOAMxfoWH8+XI+C93zgpW/UqAJo1KgCaNSoAmjUqAJo1KgCaNSoAmjUqAJo/VbQOqWoq5l7Zhv6n3eXjf7mvptau8sW6d7lzhuDr0LZRr6BxhBA6/d8wmllZJ7+ftoevtvI3jOoSa7zQjbdf//9f7wpld7Tu4vTMkB7BND6/S5HUbLutXwZe++ONzv+W6DUC11O5+vdB6B9A2i9SkuHXVlaQjNLXVez6dDP2vT4XfWzvdJG/qAota9e2jG5y7SDitKTKsal97uvyvOSSi2bjt/SrsPUQ459/35By2u3qLfVzhmYntb/Se9Xo1dS0Sd0p2sN1y3tUxS/iyr2OZemnf2/hxT1yETdWjEmo8MUdQ7taOG5zS54oErd45IWPW45CNDIt+yeziPanm4a2TT490/l0qBzHh9HVynKHdRq6l+G0HWK0o0GjfvrELrW7nHJBXrC1KeH0XhF+TldPnNiVjvHbX+ifvfc1ZMe91poGn1X3bZjjVIP+nDiRYrfRZUCGjrjRup3fEkWLVG39pn21GAaoygnsulnM6+gq+3KL6hb0R+zuwI00rSVPlAebVX30zSbnlCUDUQlykGb4wdoBv1XUarbJivKOepP1+pr6H2PSy7Qf1CUI01Tlc9oTK1q2XFbu/QzirK30/94LnOmdX9F+Q196lyjtLT0+wWXtF7tf9Ev6HqH+9k0ve6Qw3Gofjw1VVGmk+PIqHYcvf0RDTjiuG0IQCNNb1C5cvPwumuO//o7zFB7x8Xujv/xjh1z0NrX3HGpM6n/zV9CUzwuuUCrZ8jrScqttMxx4aAKuhO9ovmXF+/QP1WqNyt1f+Ug+uUe/4uOpeWOSzVZg+pAu1dQBqr7KUvplin0sXq/hQCNNN3XUVHOqx+Pdx0OOKE4L26eXXBBExVph0x1wyEa5nHJBfqMa99h9JN6u3rI8VYLaj1m1n6vZcbTktLSDenNT9StcWb1BS1X+l20N610/AgvHZZeB9q9gtJKJa8cpqGD6aB6qRKgkU/j3T8t6SnXdR9brzdNnfzv7b3Uwwgn4x9psMclF2j33Qa7QLdXrx96+cZOlLbYY51DSe51Xmv4pXAdjfK7aIp717M8/8rRAPonGjzMBfoQQCOf9pa2vr30NXqr9LDruo+t/inlius4oLXzQOML+oXHJS/Q+fSNojp3XF+1w3HE8DQN91jnKZqrftqiGq4Dfdjx/wl/i/Z0/V+jbkMD6IGkfpfLKH+q65DjE4BGvh2k+cqzCfXng/Cx1bKt4xe9NQmOS+lU5PxV8HWPS16g36Txjn3vU0Gf3dVxaTdd7rHMkJRjzs8DEirqQT9Md/hddCrNdFxa1+kOxzW7J+gH6I/qGVlo/mLnL4Un8Esh0rSENivT+tZf9bE1mUY9+Yc2WfTPo2lpqf/zwBC60q40XPICbb+Ornhocs/MFEW5mwb99Z7uNE9RMi9xfdktNMl14R/0uOuVwvvvuoY6VPpdtLIjjX6ooE2rUiWXHvEEfaIXTXroCrrKrkyhbnf9qfelAI18m5VWo4zIr7/qY+vHX3Vsee2Ghb1abk/NXnVpWp8/O34/a7jkBVqpeqBPqxv2dOzi+H3voZxm6Re/oTS8lFdc9zL3Lhro/itHYs9fVvhfVDl0R5+mHSdtVJR5uWmeKyhHf5uT2t/5wspzQ1v2uO0kQCOzJWdrL3l1sFQ9qKhOHuh548lca7+ruAugpUrM1l7y6h162PFxAXmdEfbtMdZ+V3EXQEtly9Ze8upkv+azVj6fkel1jrnOKyz+tuItgJbq9r9qL3m3/7YuTc+deiBc31B8BtCoUQXQqFEF0KhRBdCoUQXQqFEF0KhRBdCoUQXQqFEF0KhR9X9lVoORU9bMBgAAAABJRU5ErkJggg=="
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ggplot(tmp3,aes(Freq.x, Freq.y)) + geom_point() +\n",
    "    #geom_smooth(method = \"lm\", se = FALSE, color = \"black\", lwd = .2, formula = y ~ 0 + x) +\n",
    "    geom_abline(slope = 1,intercept = 0) +\n",
    "    xlab(\"# mappings, AB method\") + ylab(\"# mappings, Bowtie2\") + xlim(c(0,15)) + ylim(c(0,15))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#  check distribution of all mapping points to the tRNA locus\n",
    "distance_dist = apply(cbind(abs(untreated$pos - untreated$Start), \n",
    "                            abs(untreated$pos - untreated$End)),1,min)\n",
    "options(repr.plot.width = 8)\n",
    "options(repr.plot.height = 4)\n",
    "plot(ecdf(distance_dist), main = \"Distance of read start to mature tRNA start/stop\", \n",
    "     xlab = \"Nucleotides\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Does the number of pre-tRNA reads scale with the # of reads aligning to mature sequence only?\n",
    "\n",
    "Note that this estimate could be off because I am missing reads that map to mature tRNA sequences with modifications. To get these I would need to run ARMseq pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "all_merged = merge(tRNA_annotations, untreated, by = c(\"Chr\",\"Start\",\"End\"))\n",
    "tRNA_read_filtered = filterMature(all_merged, flip = F)\n",
    "tRNA_read_others = filterMature(all_merged, flip = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "nonpre = tRNA_read_others  %>% group_by(Anticodon) %>% do({data.frame(.$tRNA.type[1], counts = length(unique(.$qname)))})\n",
    "pre = tRNA_read_filtered  %>% group_by(Anticodon) %>% do({data.frame(.$tRNA.type[1], counts = length(unique(.$qname)))})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preANDnonpre = merge(pre,nonpre,by=\"Anticodon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ggplot(filter(preANDnonpre, counts.x < 15000), aes(x = counts.x, y = counts.y)) + \n",
    "    geom_point(aes(color = ..tRNA.type.1..x)) + \n",
    "    geom_smooth(method = \"lm\", se = FALSE, color = \"black\", lwd = .2, formula = y ~ 0 + x) +\n",
    "    xlab(\"# pre-tRNA Reads\") + ylab(\"# mature tRNA reads\") +\n",
    "    xlim(c(0,1000)) + ylim(c(0,1000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
