## Multiomic fingerprints of aging in a genetically tractable yeast population

New technologies are providing increasingly detailed measurement of biological systems. It remains a challenge, however, to integrate across these data to construct comprehensive representations of biological systems. A major goal in systems biology is to leverage heterogeneous molecular information and genetics to identify personalized intervention points that will predictively alter biological processes, such as disease. My postdoctoral project attempts to combine longitudinal multiomic profiles in a genetically diverse population with integrative computational methods to quantify genetic factors that contribute to lifespan in Baker's yeast.

Aging is a complex biological process with both genetic and non-genetic components, some of which are conserved from yeast to humans. I use multiple recently developed high-throughput technologies to image physical properties related to lifespan and quantify the molecular composition of cells at several stages of aging across two growth conditions in a genetically diverse yeast population. This large collection of longitudinal, heterogeneous data are going to be integrated using data fusion techniques and combined with genetic information using statistical machine learning methods to identify genetic loci and molecular fingerprints that are diagnostic of aging and characterize their environmental (in)dependence.

My project leverages several emerging technologies to construct a multifaceted and longitudinal survey of the aging process. Microfluidics is used to measure multiple aging phenotypes in mother cells. In parallel, high-throughput omic technologies (epigenomics, transcriptomics, metabolomics, and proteomics) are applied to quantify the molecular composition of cells at several stages of the aging process. Integration of these large heterogeneous datasets will be performed using statistical machine learning algorithms that will derive multiomic fingerprints predictive of aging state. Each of these methods will be performed in a genetically diverse yeast population of 140 haploid segregants in two growth conditions to characterize the environmental (in)dependence for each of the genetic markers and their age-specific molecular fingerprints.

Given the complexity of the aging process, I expect to observe differences in aging not only between strains with different genetic backgrounds but also within isogenic strains at different stages in their lifetimes. By combining molecular fingerprints with genetics, I will be able to identify single and combinations of genetic loci that either predict or influence lifespan. The model will make predictions about the molecular scale at which particular genetic variations manifest, from immediate consequences on transcription to downstream effects on metabolism, and will distinguish genetic factors with environmental (in)dependence. Deconstruction of this complex phenotype will help determine whether molecular profiles and genetics are sufficient to diagnose a complex biological process and provide proof-of-principle approaches for integrating large heterogeneous datasets with genetic information to drive longitudinal, data-driven evaluation of complex phenotypes, including health in humans.  

I am applying to the Joachim Herz Stiftung foundation to receive funding that will allow me to gain additional training in microfluidics, metabolomics, proteomics, and machine learning methods - each of which are central components of this project.
